packages/api/src/domain/medical/patient-demographics.ts

```diff
@@ -0,0 +1,461 @@
+    return [normalizeSsn(p.value)];
+  });
+  return removeInvalidArrayValues({
+    dob,
+    gender,
+    names,
+    addresses,
+    telephoneNumbers,
+    emails,
+    driversLicenses,
+    ssns,
+  });
+}
+
+export function removeInvalidArrayValues(demographics: LinkDemographics): LinkDemographics {
+  return {
+    dob: demographics.dob,
+    gender: demographics.gender,
+    names: demographics.names.filter(name => {
+      const nameObj: LinkGenericName = JSON.parse(name);
+      return nameObj.firstName !== "" && nameObj.lastName !== "";
+    }),
+    addresses: demographics.addresses.filter(address => {
+      const addressObj: LinkGenericAddress = JSON.parse(address);
+      return addressObj.line.length > 0 && addressObj.city !== "" && addressObj.zip !== "";
+    }),
+    telephoneNumbers: demographics.telephoneNumbers.filter(tn => tn !== ""),
+    emails: demographics.emails.filter(email => email !== ""),
+    driversLicenses: demographics.driversLicenses.filter(dl => {
+      const dlObj: LinkGenericDriversLicense = JSON.parse(dl);
+      return dlObj.value !== "" && dlObj.state !== "";
+    }),
+    ssns: demographics.ssns.filter(ssn => ssn !== ""),
+  };
+}
+
+export function normalizeDob(dob?: string): LinkDateOfBirth {
+  const parsedDate = dayjs(dob?.trim() ?? "", ISO_DATE, true);
+  if (!parsedDate.isValid()) return undefined;
+  return parsedDate.format(ISO_DATE);
+}
+
+export function normalizeGender(gender?: string): LinkGender {
+  if (gender === "M" || gender === "F") {
+    return mapGenderAtBirthToFhir(gender) as LinkGender;
+  }
+  const normalizeAndStringifydGender = gender?.trim().toLowerCase() ?? "";
+  if (normalizeAndStringifydGender !== "male" && normalizeAndStringifydGender !== "female") {
+    return undefined;
+  }
+  return normalizeAndStringifydGender;
+}
+
+export function normalizeAndStringifyNames({
+  firstName,
+  lastName,
+}: {
+  firstName: string;
+  lastName: string;
+}): string {
+  const normalizedName = {
+    firstName: firstName.trim().toLowerCase(),
+    lastName: lastName.trim().toLowerCase(),
+  };
+  return JSON.stringify(normalizedName, Object.keys(normalizedName).sort());
+}
+
+export function normalizeAddress({
+  line,
+  city,
+  state,
+  zip,
+  country,
+}: {
+  line?: string[];
+  city?: string;
+  state?: string;
+  zip?: string;
+  country?: string;
+}): LinkGenericAddress {
+  return {
+    line:
+      line
+        ?.filter(l => l !== "")
+        .map(l => {
+          return l
+            .trim()
+            .toLowerCase()
+            .replaceAll("street", "st")
+            .replaceAll("drive", "dr")
+            .replaceAll("road", "rd")
+            .replaceAll("court", "ct")
+            .replaceAll("avenue", "ave")
+            .replaceAll("lane", "ln")
+            .replaceAll("highway", "hwy")
+            .replaceAll("east", "e")
+            .replaceAll("west", "w")
+            .replaceAll("north", "n")
+            .replaceAll("south", "s");
```

Comment:
way back when when we first did the MPI, we skipped street name expanstion because we didnt have a good approach for certain edge cases: 
https://github.com/metriport/metriport/pull/1160#discussion_r1433410970
181 Court St, Brooklyn, NY 11201

I am not saying remove these, I am just re-raising this so we are aware of potential side effects. This is one of those cases where you really want a statistical approach to patient demo matching because its pretty hard to come up with a rules based heuristic to circumvent this issue

Reply:
Like you could add a rule where if there are two matches in an address, you only compress the second one, but I am sure there are edge cases there too

Reply:
discussed and settiling on using a more limited subset of replaces